delta/plugins/storage/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/storage/storages/operations/s3/S3StorageSaveFile.scala

```diff
@@ -1,76 +1,166 @@
+      },
+      _ => IO.raiseError(ResourceAlreadyExists(key))
+    )
+
+  private def convertStream(source: Source[ByteString, Any]): Stream[IO, Byte] =
+    StreamConverter(
+      source
+        .flatMapConcat(x => Source.fromIterator(() => x.iterator))
+        .mapMaterializedValue(_ => NotUsed)
+    )
+
+  private def uploadFileMultipart(fileData: Stream[IO, Byte], key: String): IO[List[Option[ETag]]] =
+    fileData
+      .through(
+        s3.uploadFileMultipart(
+          bucket,
+          FileKey(NonEmptyString.unsafeFrom(key)),
+          partSizeMB,
+          uploadEmptyFiles = true,
+          multipartETagValidation = multipartETagValidation.some
+        )
+      )
+      .compile
+      .to(List)
+
+  private def getFileAttributes(key: String): IO[GetObjectAttributesResponse] =
+    client
+      .getObjectAttributes(
+        GetObjectAttributesRequest
+          .builder()
+          .bucket(bucket.value.value)
+          .key(key)
+          .objectAttributes(ObjectAttributes.OBJECT_SIZE) // TODO get all values
+          .build()
       )
-    ).adaptError {
-      case `fileAlreadyExistException` => ResourceAlreadyExists(key)
-      case err: S3Exception            => UnexpectedSaveError(key, err.toString())
-      case err                         => UnexpectedSaveError(key, err.getMessage)
+
+  private def collectFileMetadata(
+      bytes: Stream[IO, Byte],
+      key: String,
+      uuid: UUID,
+      maybeEtags: List[Option[ETag]]
+  ): IO[FileStorageMetadata] =
+    maybeEtags.sequence match {
+      case Some(onlyPartETag :: Nil) =>
+        // TODO our tests expect specific values for digests and the only algorithm currently used is SHA-256.
+        // If we want to continue to check this, but allow for different algorithms, this needs to be abstracted
+        // in the tests and verified for specific file contents.
+        // The result will als depend on whether we use a multipart upload or a standard put object.
+        for {
+          _        <- log(key, s"Received ETag for single part upload: $onlyPartETag")
+          fileSize <- computeSize(bytes)
+          digest   <- computeDigest(bytes, storage.storageValue.algorithm)
+          metadata <- fileMetadata(key, uuid, fileSize, digest)
+        } yield metadata
+      case Some(other)               => raiseUnexpectedErr(key, s"S3 multipart upload returned multiple etags unexpectedly: $other")
+      case None                      => raiseUnexpectedErr(key, "S3 multipart upload was aborted because no data was received")
     }
+
+  // TODO issue fetching attributes when tested against localstack, only after the object is saved
+  // Verify if it's the same for real S3. Error msg: 'Could not parse XML response.'
+  // For now we just compute it manually.
+  //  private def getFileSize(key: String) =
+  //      getFileAttributes(key).flatMap { attr =>
+  //        log(key, s"File attributes from S3: $attr").as(attr.objectSize())
+  //      }
```

Comment:
The idea was both to verify the saving in S3 and using the returned attributes - object size and checksum. Currently blocked by some error. Again can be separate